### DISCUSSION TITLE
Hints / Unconstrained Input

### DISCUSSION THREAD

**Discussion by matthiasgoergens** [scroll-tech/ceno#607] [Ideas]
This is an elaboration on https://github.com/scroll-tech/ceno/discussions/551

# Guest's points of view

[Inspired by SP1](https://docs.succinct.xyz/writing-programs/inputs-and-outputs.html), we will give guests two functions to read unconstrained input (also knows as hints).  The first is `read`, which reads in a value of a given type.  The second is `read_slice`, which reads in a byte array.

```rust
pub fn read<T>() -> &'static T
where
    T: Portable + for<'a> CheckBytes<HighValidator<'a, rancor::Failure>>

pub fn read_slice() -> &'static [u8]
```

There are two main differences to SP1:
1. We are using `rkyv` instead of `bincode`.  That's because `rkyv` is a lot faster, and we are already using it in `sproll`.  The bounds on `T` in `read` [reflect that](https://docs.rs/rkyv/latest/rkyv/fn.access.html).
2. We are returning references instead of copied values.  That's because `rkyv` is zero-copy, and we want to take advantage of that.

Just like in SP1, the host has two corresponding functions to create that input: `write` and `write_slice`.  From the user's point of view, they work exactly like you imagine they would.

> Sidenote: in this model, the host has to prepare the input in exactly the order that the guest requests it.  That approach works well for simple examples.  If this inflexibility ever becomes a limiting factor, Matthias has a design to fix it.  But it's probably overkill, given that the in-order approach works well enough for SP1.

# Under the hood

Like in SP1, the typed `read` is implemented inside the guest on top of `read_slice`: it's just a wrapper that deals with deserialisation for you.  In SP1 it's [implemented exactly how you'd expect](https://docs.rs/sp1-lib/3.3.0/src/sp1_lib/io.rs.html#87).  The same for us:

```rust
pub fn read<'a, T>(&'a mut self) -> &'b T
where
    T: Portable + for<'c> CheckBytes<HighValidator<'c, Error>>,
{
    rkyv::access::<T, Error>(self.read_slice()).unwrap()
}
```

> Sidenote: users who don't like `rkyv` can wrap their own favourite deserialiser around `read_slice`.  Just like we used `rkyv` in `sproll`, even though SP1 uses `bincode` natively.

`read_slice` is a bit more involved:

At the start of the [memory-mapped region for hints](https://github.com/scroll-tech/ceno/pull/457) we store information that lets us find the byte slice for each of the items we want to read.  We keep a mutable index to the next item to read, and we increment it each time `read_slice` is called.  We then return the byte slice for that item.

> Avid readers will notice that we essentially deserialise twice.  The original design used `rkyv` for both layers, but now we code up the outer layer by hand, mostly because rkyv would prefer to pin the end of the buffer to a fixed known location, but our VM design prefers to pin the start of the buffer.

Another technical note: in general our memory-mapped-io region will be larger than the buffer `rkvy` needs for its serialised data.  In that case, we will conceptually add arbitrary padding (eg all zeroes) to the end of the buffer.

> Sidenote: SP1 mixes general RAM and memory-mapped-io.  For that reason, they need some extra system calls (like `hint_len`) to tell the difference.  We don't have that problem, so we can keep things simpler.  This also has the benefit of only using functionality we already implemented for our VM.

Our design assumes that the memory-mapped-io region is read-only.  That's a good idea for both performance and security anyway, but we can also tweak the design slightly to support read-write backing storage.

See https://github.com/scroll-tech/ceno/pull/631 for a prototype implementation.

**Comment by matthiasgoergens** [comment]
Update (will fold into the main text later):

The above is meant mostly for private IO. But the special memory region we have is for public io. So we'd want to add one for private IO, too.

We can in principle use the same idea for public input, but we probably want to Merkle-ize that by default. The guest interface might look similar, but it would be rather different under the hood.

**Comment by lispc** [comment]
looks good for me.  Just have one question: will there be a preset max total length of all unconstrained input? or as long as len(mem_map_io_region) + len(runtime_mem) <= mem_limit it is ok?

**Reply by matthiasgoergens** [reply]
Technically, yes, in this design the total length of all unconstrained input is constrained by how large that section of memory is.

Given that we have a total of 4GiB (32 bits) of address space to play with, that shouldn't be a problem in practice.

@naure You know more about how our memory works, could you clarify whether there would be any extra limits for much memory we could use here?  The mechanism outlined above just needs some read-only section of memory at a well-known location that the prover can fill in.

**Reply by naure** [reply]
It is easiest to implement it if you can reserve a fixed set of addresses - therefore a fixed maximum size. It has a cost proportional to the maximum size.

But it is also possible to implement dynamic sizes with a cost proportional to usage per execution (next power-of-two though), or dynamic addresses (chosen per execution).

**Reply by hero78119** [reply]
As non-uniform philosophy, we can define max length for `len(mem_map_io_region)`
This max can be lifted to very high value to satisfy just `len(mem_map_io_region) + len(runtime_mem) <= mem_limit`
And we dont need to worry about the overhead, current backend can support dynamic address, so we just need to pay up to `len(mem_map_io_region).next_power-of-two()` per execution

**Reply by matthiasgoergens** [reply]
Thanks.

What I would prefer for the simplest design is a fixed maximum size, say addresses `0x4000_0000..0x8000_0000` (excluding the end) to be really generous.  But then have the prover decide dynamically how much they actually want to fill up.  For technical reasons, the last address should always be fixed, the prover gets to decide the start, as in `0x8000_0000 - (1<<n) .. 0x8000_0000` in Rust range notation.

**Reply by matthiasgoergens** [reply]
To explain: You can then feed the memory region of `0x4000_0000..0x8000_0000` as a slice into `rkyv`'s `access` function, no matter how much memory you actually initialised.  Rkyv will start reading at the end, and won't care that some of the earlier memory hasn't been initialised, or might even be an error to access.

**Reply by hero78119** [reply]
What the layout of `rkyv` slices looks like when we serialized multiple (identical/diffrent) object in? I didn't find an example for that.
To expand, it will be helpful if you add this scenario in your example in https://github.com/scroll-tech/ceno/blob/matthias/ceno_host/ceno_host/src/lib.rs#L44

**Reply by matthiasgoergens** [reply]
@hero78119 Good idea, I'll flesh out the example some more.

**Reply by matthiasgoergens** [reply]
@hero78119 See `test_prepare_and_consume_items` in https://github.com/scroll-tech/ceno/blob/matthias/ceno_host/ceno_host/src/lib.rs

**Comment by kunxian-xia** [comment]
If I want to alloc `n` storage traces from hints in my guest program and `n` is not known at setup phase, then how many bytes does the slice have from calling `read_slice() -> &'static [u8]`?

**Reply by matthiasgoergens** [reply]
However many bytes are required.  Ie in the end, it's up to the prover.  But I'm not sure that answers the question you were actually aiming at?

I can code up a prototype that runs in SP1, to answer more questions, perhaps.

**Reply by hero78119** [reply]
Does the answer means in unconstrained-io/hints-inputs memory region, the memory layout is 
length_1, data, length_2, data, length_3, data, ...  ?

And this layout is determined by `write` or `write_slice` from host
So the respective `read_slice` will just read the length of bytes? 

If the answer is yes, then I think it fit with what I have done in mini-geth (former `Cannon`) and the idea is pretty straitforward.

**Reply by matthiasgoergens** [reply]
No.  The memory layout is determined by `rkyv`.  We don't need to care what it is.  (But it's different from what you described.)

You can look it up in the `rkyv` source code, if you want to, but it's basically something for the guest and host program to decide; the VM and verifier etc are agnostic about this.  And the concrete layout might change, if you upgrade to the next version of `rkyv`.

(I'm a bit blunt here: if you are curious, `rkyv` is really interesting and it's fun to read their sources.  My point is that looking this up would be just for fun and curiosity; the VM doesn't care and works no matter what rkyv does internally.)

Does this make sense?

**Reply by hero78119** [reply]
ok I got it, so `rkyv` crate encode the serialized format and layout for us . So I think the answer to @kunxian-xia question is: the bytes `n` will be determined per proof, when host execute the program first to prepare private io data.

**Reply by naure** [reply]
I think Kunxian you’re asking about the underlying segment reserved for private input. That can be a `&[u8]` with a constant `len` inside, or a `[u8; LEN]`. That’s the fixed maximum size.

From there, the guest/host have to agree on a data format; if that data contains some variable-size structures (e.g. `Vec`) they will encode their own sizes somehow (e.g. however `rkyv` does it).

**Reply by matthiasgoergens** [reply]
Looks like you are getting at what I am describing, despite my probably very confusing wording.

@naure You are really good with words, I love your PR descriptions.  Feel free to improve the write-up at the top of this page, please!  (Everyone else: also feel free to make improvements, please.)

**Reply by kunxian-xia** [reply]
@matthiasgoergens ~~Do you mean just let rkyv to decide how many bytes it should return?~~ 
The host / prover will execute the program before proving to get how many bytes are needed and then use that data to init memory region? Then my question becomes:
> let's say the guest program invokes `read_slice()` by three times? Where are these three fat pointers `&[u8]` stored in memory?

**Reply by matthiasgoergens** [reply]
Simplified: the three pointers are stored in the memory mapped private IO region.

Accurate: the pointers are calculated from information stored in the memory mapped private IO region.  The main difference is that rkyv stores relative pointers there, and they are converted to absolute pointers on access.

> The host / prover will execute the program before proving to get how many bytes are needed and then use that data to init memory region?

More or less, yes.  But the host doesn't get the information from executing the guest program in this simplified model.  It's just programmed with that knowledge already.  See how https://github.com/scroll-tech/sproll-evm/blob/bd1b3b118be28d9b1bfbd544db60da7ef1e21223/host/src/runner.rs#L85 knows how many times to run `write_slice` in a loop, because it's aware of the logic that the guest program uses.

To quote myself:

> Sidenote: in this model, the host has to prepare the input in exactly the order that the guest requests it. That approach works well for simple examples. If this inflexibility ever becomes a limiting factor, Matthias has a design to fix it. But it's probably overkill, given that the in-order approach works well enough for SP1.

**Reply by matthiasgoergens** [reply]
@kunxian-xia Based on feedback, I changed the layout.  Now we (more-or-less) at the start of the hints memory region we store explicit pointers to the different items.  Please see above.

**Comment by hero78119** [comment]
Have a question regarding e.g. read number of round in `Fibonacchi` and what happened under the hood, in [inspired by sp1](https://docs.succinct.xyz/writing-programs/inputs-and-outputs.html#reading-data)

```rust
let round = sp1_zkvm::io::read::<u32>();
sp1_zkvm::io::commit::<u32>(&round);
```

Does it means that `round` u32 data will appear in 2 place, one is public-input memory region, another is hint-data memory region?
Or if the commit call is happened, the data will be move to public-input (constrained data) memory region only?

**Reply by matthiasgoergens** [reply]
That's an exceedingly good question!  For now, I only address the `read` part here.  Let me try and whip up a design for the `commit` part next.

I suspect both the approaches you outlined are probably feasible (and a few more other approaches are feasible, too).  Let me spend a bit of time thinking through them and work out the consequences.

**Reply by matthiasgoergens** [reply]
@hero78119 https://github.com/scroll-tech/ceno/discussions/627 is the sketch for public input.  The suggestion there avoids the need for the committed values to appear multiple times.

**Comment by matthiasgoergens** [comment]
@kunxian-xia and me had a fruitful discussion about the approach outlined above that uses a prover initialised read-only memory region versus SP1's approach that mixes hints in with general read-write memory, but requires two system calls (`hint_len` and `hint_read`) to coordinate with the prover.

In SP1's system, if you want a 'hint' you first call `hint_len` to get the length of the hint, than you get memory of that length from the general purpose memory allocator.  Crucially, that memory is uninitialised, as far as Rust is concerned.  But we use some `unsafe` tags to tell the Rust compiler not to worry.  Crucially, the prover decides what values unitialised memory starts with.  So this way, the prover can pass you some private inputs.  You then feed the slice of freshly allocated memory to the deserialiser of your choice.  SP1 suggests bincode, but anything works.

Well, strictly speaking what I described above is enough to make the scheme work.  But to make coordination between guest and host program simpler, SP1 has the `hint_read` system call that you can use to tell the prover where to write the hint.  But there are no constraints associated with it as far as I can tell.

(A host program that knows its guest program well enough, could do without `hint_read` and just _know_ how to initialise the memory.  But that would be really annoying to program for.)

Our VM is different: our circuits guarantee that general read-write memory that our Rust allocator has access to comes zeroed out from the factory.  On the downside, that means we need an extra memory region for the prover to leave hints in.  But we get (at least) two upsides:
- Getting zero initialised memory from the allocator is [essentially free](https://github.com/scroll-tech/ceno/pull/336).  The Rust compiler is smart enough to make use of that for examples like `let foo = vec![0_u32; 128 * 1024];` (but more practically releant for eg setting up hash tables).
- In our VM design, accessing read-only memory is cheaper for the prover than accessing read-write memory.

**Reply by naure** [reply]
Some corrections here.

- Memory can be initialized to any content that we want. Constant 0 is a special case of it. We can lift that constraint to allow prover hints in https://github.com/scroll-tech/ceno/pull/617.

- Memory is not free, it costs something for a slot to be available even it is not actually accessed.

- There are no implementations of read-only memory at the moment. That can be added and it is plausibly more efficient, but the cost analysis is a little complex because it requires some kind of permission mechanism too.

**Reply by hero78119** [reply]
(just leave the comment before noticing Aurel has replied earlier 😂 )

> Cite from @kunxian-xia claim: ...With the hint inputs stored in a static region approach, the verifier has to make sure the “static” memory region for hint inputs is disjoint from the “dynamic” memory region used by heap.

Right now in ceno backend memory scheme, we support partition overall memory space into "arbitrary" regions, and within each region, the real in use memory per execution can be dynamic stretching. So it's not an issue since verifier just assure stretching not exceeding in `max_length` for each region. So I would say we dont need to worry about that and force us to chose mixed hints data with global allocator region.

Regarding to cost, read-only is slightly cheaper as we dont need to commit "final_v" column. Other cost for both type are almost identical.

Our memory scheme is pretty flexible, so I think separate hint data with allocator (heap) make sense to me, at least we dont need to have  "hint_len, hint_read" sys call, so the overall execution trace (cycles) can be shorter. 

I am wondering, is there any other significant benefit to have "hint_len, hint_read" ? e.g. it makes guest/host execution flow more identical or sth...

**Reply by kunxian-xia** [reply]
> So it's not an issue since verifier just assure stretching not exceeding in max_length for each region.

The verifier needs to make sure regions are disjoint, right?

**Reply by kunxian-xia** [reply]
> is there any other significant benefit to have "hint_len, hint_read" ?

The cons of using static region is we need to pay overhead even for untouched memory slot. For example, if the static region has 16MB, then we need to use a multilinear polynomial with 24 variables to store the `init_value`. If a concrete prover instance only has 1MB hint inputs, the prover's overhead is 15x of the actual need. If we just treat hint_inputs as normal object allocated in heap, then we can avoid this additional overhead.

**Reply by kunxian-xia** [reply]
> we dont need to have "hint_len, hint_read" sys call

The number of cycles to execute `hint_len / hint_read` is really neglible as it only need write at most three registers.
```rust
// https://github.com/succinctlabs/sp1/blob/dev/crates/zkvm/entrypoint/src/syscalls/io.rs#L46-L50
// hint_len
asm!(
    "ecall",
    in("t0") crate::syscalls::HINT_LEN,
    lateout("t0") len,
);
// https://github.com/succinctlabs/sp1/blob/dev/crates/zkvm/entrypoint/src/syscalls/io.rs#L64-L69
// hint_read
asm!(
    "ecall",
    in("t0") crate::syscalls::HINT_READ,
    in("a0") ptr,
    in("a1") len,
);
```

**Reply by hero78119** [reply]
> The verifier needs to make sure regions are disjoint, right?

yes and we already implemented it in verifier, which just one assertion.
https://github.com/scroll-tech/ceno/blob/12ed92231220be35574eb176d454be0324825f56/ceno_zkvm/src/scheme/verifier.rs#L546

> The cons of using static region is we need to pay overhead even for untouched memory slot. For example, if the static region has 16MB

No need. As Matt point out earlier, here the concept are mixed up.

In summary: we already implemented and support NON static regions even no `hint_len` and `hint_read`, we can just based on `read/read_slices` to achieve the ideal situation you have in mind.

**Reply by kunxian-xia** [reply]
> just one assertion.

The code you link here checks "verifier just assure stretching not exceeding in max_length for each region". It seems the verifier has not check all memory regions are disjoint, right?

**Reply by kunxian-xia** [reply]
If memory regions are not disjoint, it's possible that ram circuits will generate two write records with same address and `cycle = 0` `(a, v, 0), (a, v', 0)`. Then the prover can manipulate the value that first load instruction gets (either `v` or `v'` works).

**Reply by hero78119** [reply]
> The code you link here checks "verifier just assure stretching not exceeding in max_length for each region". It seems the verifier has not check all memory regions are disjoint, right?

Disjoin check are achieved by
- [ ] in "setup" phase, check each memory region offset + max_len not overlapping => haven't done, but easy, no verifier cost
- [x] in verifier per proof, check hint_vars not exceed max_len by just one assertion

**Reply by kunxian-xia** [reply]
> here the concept are mixed up.

I see.

**Reply by naure** [reply]
> The verifier needs to make sure regions are disjoint, right?

Yes. It looks like this right now:
https://github.com/scroll-tech/ceno/blob/17a9d254d1453108603aaa6da06ed6db40a91013/ceno_zkvm/src/instructions/riscv/rv32im/mmu.rs#L56

**Reply by matthiasgoergens** [reply]
> The number of cycles to execute hint_len / hint_read is really neglible as it only need write at most three registers.

Yes, it's not about the cycles, is about having to implement yet another two syscalls.  That adds to complexity of the design and increases attack surface.

**Comment by matthiasgoergens** [comment]
See https://github.com/scroll-tech/ceno/discussions/627 for public input / 'commit'.

**Comment by matthiasgoergens** [comment]
@hero78119 See https://github.com/scroll-tech/ceno/pull/631 for the same code that's on the branch, but easier to view all the changes at once.
